### PR TITLE
Adding default config directories.

### LINKED ISSUE #10338: default profile directory
https://github.com/servo/servo/pull/10114 introduces a preference file.

The profiler directory is set via the `--profile-dir` option.

We should have a default directory.

This directory should not be created if it doesn't exist, but just if `--profile-dir` is not specified and `default_profile_dir/prefs.json` exists, then we should read this file.

I guess these are good-enough values:
- windows: `%APPDATA%\servo\`
- mac: `~/Library/Application Support/Servo/`
- linux: use XDG Base Directory spec (or just `~/.config/servo/`)

I would not be surprised if there was already a crate that resolves the default configuration directory for each platform.

### PR DESCRIPTION
**Pull request by creativcoder**
Fixes #10338 

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10498)

### CODE REVIEW

File: components/servo/Cargo.lock [hunk 1]
```diff
@@ -2214,6 +2216,15 @@ dependencies = [
 
 [[package]]
 name = "uuid"
+version = "0.1.18"
```
Comment (Manishearth) [inline components/servo/Cargo.lock]: We shouldn't depend on two versions of the same crate if we can help it.

Comment (creativcoder) [inline components/servo/Cargo.lock]: Ok. So 0.1.18, needs to be removed and uuid 0.2.0 needs to stay; right ?

**Comment by KiChjang** [comment]
@creativcoder Also, just a small nit, in your PRs, please say something like "fixes #10338" or "addresses #10338" or "closes #10338" instead of "for #10338". This will cause GitHub to automatically close the issue number you stated once the PR lands.

File: components/util/Cargo.toml [hunk 2]
```diff
@@ -45,7 +45,9 @@ serde = "0.7"
 serde_macros = "0.7"
 smallvec = "0.1"
 string_cache = {version = "0.2.11", features = ["heap_size"]}
-url = {version = "0.5.7", features = ["heap_size", "serde_serialization"]}
+url = {version = "0.5.8", features = ["heap_size", "serde_serialization"]}
+uuid = "0.1.17"
```
Comment (Ms2ger) [inline components/util/Cargo.toml]: I removed the dependency on uuid here, please don't add it back.

**Comment by aneeshusa** [comment]
This should also remove the `--profile-dir` option, since it's being replaced by `--config-dir`.

**Comment by creativcoder** [comment]
Ok. I assumed that `config-dir` was an additive option. I'll make the necessary changes then.

**Comment by SimonSapin** [comment]
Reviewed 4 of 4 files at r3, 2 of 3 files at r4.
Review status: 6 of 7 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[components/util/basedir.rs, line 18 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10498#-KF5pfuIXcRuK3y5AWD9:-KF5pfuJ8xWSUI6bFEsI:1296764573) ([raw file](https://github.com/servo/servo/blob/99107f305b406ae8565ab369402aee3dcfaec61e/components/util/basedir.rs#L18)):*
Shouldn’t this and other functions in this file be based on the `--config-dir` command-line option? We don’t want to touch the default location if another one is specified.

---

*[components/util/basedir.rs, line 21 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10498#-KF5q5SJR3URyH4UGClA:-KF5q5SK-lioz2NzQNk8:-1396545213) ([raw file](https://github.com/servo/servo/blob/99107f305b406ae8565ab369402aee3dcfaec61e/components/util/basedir.rs#L21)):*
Shouldn’t this use `with_profile`?

---

*[components/util/basedir.rs, line 53 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10498#-KF5qVt6dPkdmor6q5v3:-KF5qVt7QYytezkwlusC:-993040544) ([raw file](https://github.com/servo/servo/blob/99107f305b406ae8565ab369402aee3dcfaec61e/components/util/basedir.rs#L53)):*
Libraries in general should not swallow errors or print to `stderr`. Instead return a `Result` and let the caller decide how to deal with errors.

---

*[components/util/basedir.rs, line 59 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10498#-KF5qogOZVdNiqq3XeJe:-KF5qogOZVdNiqq3XeJf:-749719954) ([raw file](https://github.com/servo/servo/blob/99107f305b406ae8565ab369402aee3dcfaec61e/components/util/basedir.rs#L59)):*
Is this line unused?

---

*[tests/unit/util/prefs.rs, line 50 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10498#-KF5ookYd_z_Nm041zmR:-KF5ookYd_z_Nm041zmS:102678145) ([raw file](https://github.com/servo/servo/blob/99107f305b406ae8565ab369402aee3dcfaec61e/tests/unit/util/prefs.rs#L50)):*
Use `let _ = something_that_returns_result;` instead of this to ignore errors. But tests should generally not ignore errors, so use `something_that_returns_result.unwrap();` instead.

---

*[tests/unit/util/prefs.rs, line 63 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10498#-KF5oX7PNtqD3Cma98Lo:-KF5oX7QUHchI4KHQUbP:870994287) ([raw file](https://github.com/servo/servo/blob/99107f305b406ae8565ab369402aee3dcfaec61e/tests/unit/util/prefs.rs#L63)):*
Nit: why the leading `_`? That’s usually used to suppress "unused variable" warnings, but this *is* used here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10498)*

**Comment by creativcoder** [comment]
*[components/util/basedir.rs, line 18 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10498#-KF5pfuIXcRuK3y5AWD9:-KFArwbm9cES00_Eqve2:-208519177) ([raw file](https://github.com/servo/servo/blob/99107f305b406ae8565ab369402aee3dcfaec61e/components/util/basedir.rs#L18)):*
Can you elaborate on this one ? I am not clear on the intention here. As for using these functions; only the `bootstrap_default_dirs` is being called, when setting the initial environment options ( method `from_cmdline_args` in `opts.rs`) for servo, the other functions are meant to be called only ( in `prefs.rs` ) when checking if this flag is present there. Correct me if that is not the right place to call, `bootstrap_default_dirs`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10498)*

**Comment by SimonSapin** [comment]
Review status: 0 of 10 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[components/util/basedir.rs, line 18 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10498#-KF5pfuIXcRuK3y5AWD9:-KGHaQBvUvzIoB86XK7W:-2055559099) ([raw file](https://github.com/servo/servo/blob/99107f305b406ae8565ab369402aee3dcfaec61e/components/util/basedir.rs#L18)):*
My point is: why does `bootstrap_default_dirs` exist at all? If I’m running `servo --config-dir ~/path/to/config/servo` then `~/.config/servo` should not be created at all.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10498)*

**Comment by paulrouget** [comment]
@creativcoder any update on this?

**Comment by creativcoder** [comment]
Hi @paulrouget ! I had made changes a week ago. Will rebase and update soon.

**Comment by SimonSapin** [comment]
Looks good, thanks!

@bors-servo r+

**Comment by SimonSapin** [comment]
@creativcoder This needs you to run `./mach cargo-update -p util` and commit the changes.

**Pull request merged**